Hookup GraphQL models in resolvers

jeffreyzhang2001 opened the pull request:
## Notion ticket link
<!-- Please replace with your ticket's URL -->
[Hook up GraphQL resolvers together for various models](https://www.notion.so/uwblueprintexecs/ac7b1f636a214a0cbf45f6b76f899b54?v=5414ee5fc5974f149921e192d6eb29b8&p=a4cbc48f95734f8b9a49f9ee01845504)


<!-- Give a quick summary of the implementation details, provide design justifications if necessary -->
## Implementation description
* Added types for Guardian and MedicalInformation
* Added ability to query employees by ID (for future usage)
* Added eager loading on a bunch of queries with Prisma includes (e.g. fetch applicants' applications when querying for applicants)

<!-- Catch all section that could be used to draw attention to anything the reviewers should keep in mind, substantial parts of your PR, anything you'd like a second opinion on, things that will be fixed in the future, or things that were left out -->
## Notes
* Please look through the PR and let me know if there's anything I missed :D


## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x ] I have run the appropriate linter(s)
- [x ] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

angeladietz commented on lib/medicalInformation/schema.graphql:
```diff
@@ -0,0 +1,12 @@
+type MedicalInformation {
```
Do we also need to be adding resolvers for the `MedicalInformation` and `Guardian`?

OustanDing replied:
Not at the top level. However, we would want to have field resolvers for medical information and guardian so that they can be retrieved when querying for applicants. I mentioned this in my comment above.

angeladietz commented on lib/applicants/schema.graphql:
```diff
@@ -14,11 +14,11 @@ type Applicant {
   address: String!
   postalCode: String!
   rcdUserId: Int
-  # status: ApplicantStatus
```
You can also remove the TODO comment at the beginning of this file

angeladietz commented on lib/applications/resolvers.ts:
```diff
@@ -28,7 +32,12 @@ export const createApplication: Resolver = async (_, args, { prisma }) => {
+      data: {
+        ...args.input,
+        applicant: {
+          connect: { id: applicantId },
```
Should we also be doing a `connect` like this in `lib/permits/resolvers`? 

Also, there's some TODOs in `lib/permits/schema.graphql` regarding the same task as this PR. I'm not totally sure if your changes already fix them or not but I think it would make sense for you to address them in this pr.

OustanDing replied:
Yes we should have a `connect` when creating permits. We always want a created permit to be attached to an applicant.

angeladietz reviewed (dismissed):
Looks great overall! Just left a few small comments.

OustanDing commented on lib/applicants/resolvers.ts:
```diff
@@ -9,7 +9,14 @@ import { formatPhoneNumber, formatPostalCode } from '@lib/utils/format';
  */
 export const applicants: Resolver = async (_parent, _args, { prisma }) => {
-  const applicants = await prisma.applicant.findMany();
+  const applicants = await prisma.applicant.findMany({
+    include: {
+      applications: true,
+      permits: true,
+      guardian: true,
+      medicalInformation: true,
+    },
+  });
```
We don't want to include all of these fields off the bat. The goal of GraphQL is to only do as much work as needed, so we don't want to do all these joins and queries if we're not asking for these fields.

Instead, we want to make use of field resolvers to retrieve these nested fields, like so (from Ben Awad video lol):
![image](https://user-images.githubusercontent.com/30249230/125838843-28a769ff-9c18-4ea4-b9f1-c587711eb3a6.png)

We'll address N+1 problem later.

OustanDing commented on lib/applications/resolvers.ts:
```diff
@@ -12,7 +12,11 @@ import { DBErrorCode } from '@lib/db/errors'; // Database errors
  */
 export const applications: Resolver = async (_parent, _args, { prisma }) => {
-  const applications = await prisma.application.findMany();
+  const applications = await prisma.application.findMany({
+    include: {
+      permits: true,
+    },
+  });
```
Same here as above.

OustanDing commented on lib/guardian/schema.graphql:
```diff
@@ -0,0 +1,14 @@
+  firstName: String!
+  middleName: String
+  lastName: String!
+  address: String!
```
Good heads up. We probably just want to do change `address` to `addressLine1` and `addressLine2` in this PR then. I don't think it will raise conflicts.

OustanDing replied:
Actually, I think we can still merge #24 first, then rebase. This PR doesn't have any database changes, and the GraphQL schema changes depend on the SQL schemas.

OustanDing commented on lib/physicians/resolvers.ts:
```diff
@@ -10,7 +10,11 @@ import { formatPhoneNumber, formatPostalCode } from '@lib/utils/format';
  */
 export const physicians: Resolver = async (_parent, _args, { prisma }) => {
-  const physicians = await prisma.physician.findMany();
+  const physicians = await prisma.physician.findMany({
+    include: {
+      medicalInformation: true,
+    },
+  });
```
Same comment as above.

jeffreyzhang2001 commented:
Addressed all the comments, I know it's a bit of a big PR but there are a bunch of patterns across the changes so it shouldn't be too bad.

Note: You can re-seed your database with the seed script included in this PR (`npx prisma db seed --preview-feature`)

OustanDing commented on lib/applications/field-resolvers.ts:
```diff
@@ -0,0 +1,18 @@
+  _args,
+  { prisma }
+) => {
+  return await prisma.permit.findMany({ where: { applicationId: parent?.id } });
```
This should be `findUnique` since every application can only be related to at most one permit.

jeffreyzhang2001 replied:
The way our schema is defined, the `Application` model has multiple `Permit` objects associated with it. To fix this, we would have to modify our schema and change a bunch of types. For sake of getting an MVP out, I think it'll be fine to leave as is and fix later - `findMany` with the ID restriction will still work as desired, since each `Permit` is only associated with one object.

OustanDing replied:
Good point. Nice catch! I will add a ticket to the backlog to address this post-MVP.

OustanDing commented on lib/employees/resolvers.ts:
```diff
@@ -4,8 +4,8 @@ import { EmployeeAlreadyExistsError } from '@lib/employees/errors'; // Employee
 /**
- * Query all the RCD employees in the internal-facing app
- * @returns All RCD employees
+ * Queries for RCD employees in the internal-facing app
+ * @returns An array of RCD employees
```
We can revert this comment.

OustanDing commented on lib/guardian/schema.graphql:
```diff
@@ -0,0 +1,15 @@
+  middleName: String
+  lastName: String!
+  addressLine1: String!
+  addressLine2: String!
```
```suggestion
  addressLine2: String
```

OustanDing commented on lib/guardian/schema.graphql:
```diff
@@ -0,0 +1,15 @@
+  phone: String!
+  relationship: String!
+  notes: String
+  applicant: Applicant!
```
I don't think this should be included in the schema, since we don't have a field resolver for `Applicant` under `Guardian`. Also, we wouldn't be querying for that anyway.

OustanDing commented on lib/permits/resolvers.ts:
```diff
@@ -21,30 +21,42 @@ export const permits: Resolver = async (_parent, _args, { prisma }) => {
   try {
     permit = await prisma.permit.create({
-      data: { ...input },
+      data: {
+        ...args.input,
+        applicant: {
+          connect: { id: applicantId },
+        },
+        application: {
+          connect: { id: applicationId },
+        },
+      },
```
Nice :D

OustanDing commented on prisma/seed.ts:
```diff
@@ -1,14 +1,18 @@
 import prisma from '../prisma/index'; // Relative path required, path aliases throw error with seed command
-import { Role } from '@prisma/client';
-
+import { Role, Gender, Province, PaymentType, Aid } from '../lib/graphql/types';
+/**
+ * Main program
+ */
```
```suggestion
import { Role, Gender, Province, PaymentType, Aid } from '../lib/graphql/types'; // Relative path required

/**
 * Main program
 */
```

OustanDing reviewed (changes requested):
Mostly documentation-related comments. The field resolvers look great and make sense! After resolving the issues, we should be good to go.

OustanDing reviewed (approved):
LGTM! Thank you for addressing the PR comments quickly!
